vyper/semantics/validation/data_positions.py

```diff
@@ -18,7 +20,125 @@ def set_data_positions(vyper_module: vy_ast.Module) -> StorageLayout:
         Top-level Vyper AST node that has already been annotated with type data.
     """
     set_code_offsets(vyper_module)
-    return set_storage_slots(vyper_module)
+    return (
+        set_storage_slots_with_overrides(vyper_module, storage_layout_overrides)
+        if storage_layout_overrides
+        else set_storage_slots(vyper_module)
+    )
+
+
+class StorageCollision:
+    """
+    Keep track of which storage slots have been used. If there is a collision of
+    storage slots, this will raise an error and fail to compile
+    """
+
+    def __init__(self):
+        self.occupied_slots: Dict[int, bool] = {}
+
+    def is_slot_free(self, slot_number: int) -> bool:
+        return not self.occupied_slots.get(slot_number)
+
+    def are_all_slots_free(self, slots_to_check: List[int]) -> bool:
+        return not any(self.occupied_slots.get(slot) for slot in slots_to_check)
+
+    def reserve_slot(self, slot_number: int) -> None:
+        self.occupied_slots[slot_number] = True
+
+    def reserve_slots(self, slots_to_reserve: List[int]) -> None:
+        for slot in slots_to_reserve:
+            self.occupied_slots[slot] = True
+
+    def check_and_reserve_slot(self, slot_number: int) -> None:
+        if self.occupied_slots.get(slot_number):
```

Comment:
better to use the API that's already set up -
```suggestion
        if not self.is_slot_free(slot_number):
```